File: 2022/Day 2/pt2.cpp
#include <helper.hpp>
#include <map>

std::map<char, std::map<char, char>> decision_map;
std::map<char, std::map<char, int>> round_score_map;
std::map<char, int> choice_score_map{{'A', 1}, {'B', 2}, {'C', 3}};

void populate_round_score_map()
{
    round_score_map['A'] = {{'A', 3}, {'B', 6}, {'C', 0}};
    round_score_map['B'] = {{'A', 0}, {'B', 3}, {'C', 6}};
    round_score_map['C'] = {{'A', 6}, {'B', 0}, {'C', 3}};
}

void populate_decision_map()
{
    decision_map['A'] = {{'X', 'C'}, {'Y', 'A'}, {'Z', 'B'}};
    decision_map['B'] = {{'X', 'A'}, {'Y', 'B'}, {'Z', 'C'}};
    decision_map['C'] = {{'X', 'B'}, {'Y', 'C'}, {'Z', 'A'}};
}

int get_total_score(std::vector<std::string> lines)
{
    int total_score = 0;

    for (std::string line : lines)
    {
        char opnt, your;
        sscanf(line.c_str(), "%c %c", &opnt, &your);
        your = decision_map[opnt][your];

        total_score += round_score_map[opnt][your] + choice_score_map[your];
    }

    return total_score;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");

    auto t1 = Helper::get_current_time();

    populate_round_score_map();
    populate_decision_map();
    int score = get_total_score(lines);
    std::cout << "Output: " << score << std::endl;

    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/Day 01/pt2.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/benjababe/advent-of-code/helper"
)

func findFirst(numMap map[string]string, line string) int {
	for len(line) > 0 {
		for num := int64(1); num < 10; num++ {
			if string(line[0]) == strconv.FormatInt(num, 10) {
				return int(num)
			}
		}

		for numText, num := range numMap {
			if strings.HasPrefix(line, numText) {
				i, _ := strconv.Atoi(num)
				return i
			}
		}

		line = line[1:]
	}

	return -1
}

func findLast(numMap map[string]string, line string) int {
	for len(line) > 0 {
		lastI := len(line) - 1

		for num := int64(1); num < 10; num++ {
			if string(line[lastI]) == strconv.FormatInt(num, 10) {
				return int(num)
			}
		}

		for numText, num := range numMap {
			if strings.HasSuffix(line, numText) {
				i, _ := strconv.Atoi(num)
				return i
			}
		}

		line = line[:lastI]
	}

	return -1
}

func solve(lines []string) int {
	score := 0

	numMap := map[string]string{
		"one": "1", "two": "2", "three": "3",
		"four": "4", "five": "5", "six": "6",
		"seven": "7", "eight": "8", "nine": "9",
	}

	for _, line := range lines {
		first := findFirst(numMap, line)
		last := findLast(numMap, line)
		score += first*10 + last
	}

	return score
}

func main() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solve(lines)
	fmt.Printf("Output: %d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)
}
